pkg/webhooks/machine_webhook.go

```diff
@@ -103,9 +111,28 @@ var (
 			maxProcessor:             143,
 		},
 	}
+
+	cloudProviderArchMachineTypes = map[osconfigv1.PlatformType]map[machineArch]string{
+		osconfigv1.AWSPlatformType: {
+			AMD64: defaultAWSX86InstanceType,
+			ARM64: defaultAWSARMInstanceType,
+		},
+		osconfigv1.AzurePlatformType: {
+			AMD64: defaultAzureX86VMSize,
+			ARM64: defaultAzureARMVMSize,
+		},
+		osconfigv1.GCPPlatformType: {
+			AMD64: defaultGCPX86MachineType,
+			ARM64: defaultGCPARMMachineType,
+		},
+	}
 )
 
 const (
+	arch              = machineArch(goruntime.GOARCH)
+	ARM64 machineArch = "arm64"
+	AMD64 machineArch = "amd64"
```

Comment:
How will this work when we have heterogeneous clusters?

What happens if the arch is not one of these two? I think we need to have a fallback because the tests running on a macbook will not pass given this, they will find an empty value from the map lookup

Reply:
Hi, I changed the code to use a function; it returns the amd64 default instance type for the given provider if failing the lookup. It still returns an empty string if the function is used for a non-supported provider. Is that ok?

I also moved the map to be local to the function for "disallowing modifications", if that is worth it.

> How will this work when we have heterogeneous clusters?

In the case of heterogeneous clusters, we default to the instance type of the same architecture as the control plane. This is the same behavior previously implemented for AWS only. I don't see other ways to define a 'primary architecture' and its default instance type here.


Once the review is done and we agree on an approach, I'll squash the commits properly.

Reply:
Rebased and added logs and warnings